File: Week 2/2_last_digit_of_fibonacci_number/fibonacci_last_digit.py
# Uses python3
def get_fibonacci_last_digit_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % 10


def get_fib_fast(n):
    fib = [-1 for x in range(n+2)]
    fib[0] = 0
    fib[1] = 1

    if fib[n] != -1:
        return sum([fib[x] for x in range(n + 1)]) % 10


    else:
        for i in range(2, n+1):
            fib[i] = (fib[i-1] % 10 + fib[i-2] % 10) % 10
        return sum([fib[x] % 10 for x in range(n + 1)]) % 10


if __name__ == '__main__':
    n = int(input())
    for i in range(n):
        print(get_fib_fast(i))

File: Week 2/2_last_digit_of_fibonacci_number/fibonacci_last_digit.cpp
#include <bits/stdc++.h>
using namespace std;
// int get_fibonacci_last_digit_naive(int n) {
//     if (n <= 1)
//         return n;

//     int previous = 0;
//     int current  = 1;

//     for (int i = 0; i < n - 1; ++i) {
//         int tmp_previous = previous;
//         previous = current;
//         current = (tmp_previous%10 + current%10) % 10;
//     }

//     return current % 10;
// }

int get_fib_fast(int n) {
    vector<int> fib(n+1, -1);
    fib[0] = 0;
    fib[1] = 1;

    if(fib[n] != -1)
        return fib[n];
    
    else {
        for(int i = 2; i <= n; i++)
            fib[i] = (fib[i-1] % 10 + fib[i-2] % 10) % 10;
        return fib[n];
    }
}

int main() {
    int n;
    std::cin >> n;
    // int c = get_fibonacci_last_digit_naive(n);
    std::cout << get_fib_fast(n) << '\n';
    }
